rfcs/0019-maintainers.md

```diff
@@ -0,0 +1,129 @@
+To enable this, we must be able to map all nixpkgs files to maintainers.
+
+
+# Motivation
+[motivation]: #motivation
+<!--  Why are we doing this? -->
+Currently we have two mechanism in place that explictly describes who the owner is.
+It either is defined in the `meta.maintainers` in a `stdenv.mkDerivation`, or when it matches a
+pattern in `CODEOWNERS` file.
+
+The `meta.maintainers` only covers the packages, and is not used to determine reviewers
+for packages on GitHub.
+
+The [`CODEOWNERS`](https://help.github.com/articles/about-codeowners/)
+is only used to help GitHub select a reviewer and is centrally managed in a single file.
+One of the problems is that this loses the data locality, unlike the `.gitignore` files, on which
+the format is heavily inspired.
+This makes further delegation of maintainer responsibilities harder.
+
+These two systems do not integrate together.
+The meta data in `pkgs/*` is used by Hydra packages break, and the CODEOWNERS is only used by
+GitHub for reviews.
+
+The `CODEOWNERS` format is not extensible at all; there is no way to specify e.g. that some parts
+require one positive review of the maintainers and others might require more than one.
+
+Furthermore, GitHub is not suitable to handle the permission that the Nix community needs;
+currently people either have full commit access, or none at all.
+
+Therefore, we should implement a format and automation to handle both the data from the
+`meta.maintainers` and some maintainers file.
+
+The output of this RFC is:
+- An agreed format for the maintainer file format.
+- A simple tool (with machine parseable output) to answer who is the maintainer
+  of a file or git diff.
+- Have a check to ensure that every file in nixpkgs has a maintainer.
+
+
+# Detailed design
+[design]: #detailed-design
+
+<!-- This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with the ecosystem to understand, and implement.  This should get
+into specifics and corner-cases, and include examples of how the feature is
+used. -->
+
+## Requirements
+The input to the maintainers script is a list of files the output will be a machine readable list
+of maintainers. It will support a `--why` option to print out which decisions points were
+encountered, in order to debug why someone is a maintainer of a file.
+
+This script will read in a MAINTAINERS files, with support for nested delegation like how the
+`.gitignore` files work.
+
+For files in `pkgs/`, the script will try to use the metadata in the packages to find the maintainer.
+If this information cannot be retrieved, it falls back to the rules in the MAINTAINERS file.
+
+
+## Maintainers file format
+
+TBD.
+
+There are two high level options to take:
+- Data format, such as toml, yaml, or the
+  [Linux kernel's maintainer file format](https://github.com/torvalds/linux/blob/master/MAINTAINERS)
+- Use Nix file + script to invoke this.
```

Comment:
This seems like a good option, since then we can include lib/maintainers.nix from that file. And it can easily be converted to JSON for processing by other tools (`nix eval --json`). The format could be something like:
```
with import lib/maintainers.nix;

[
  { description = "Stdenv";
    paths = [ "pkgs/stdenv/*" ... ];
    maintainers = [ eelco ... ];
  }

  { description = "Haskell Infrastructure";
    paths = [ "pkgs/top-level/haskell-packages.nix"
              "pkgs/development/compilers/ghc/*"
              "pkgs/development/haskell-modules/lib.nix"
            ];
    maintainers = [ peti ... ];
  } 
]
```